### DISCUSSION TITLE
How to make sql_exporter consider different database versions when fetching data?

### DISCUSSION THREAD

**Discussion by blueeye2015** [burningalchemist/sql_exporter#473] [Q&A] [answered]
When I use the exporter fetching data from Mysql's information schema tables , different version of databases have different metrics table name.  
How can I make sql_exporter execute different SQL queries to fetch data depending on the version?

**Comment by burningalchemist** [comment, marked as answer]
Hi @blueeye2015, there is no way to set conditionals for queries, as we would need an additional query to identify the version of the database. And each database has a different way to store the version. It doesn't play well with the agnostic nature of this exporter.

So currently the config file might need to be pre-rendered based on the requirements. It happens outside of sql_exporter anyways.

Other available options are (in the order of increasing complexity):
 - have two sql_exporters with two separate queries;
 - have two separate queries within a single exporter, ignore errors;
 - have two separate jobs in `sql_exporter` and call them separately with query parameter `/metrics?jobs[]=job_name`;
 - adjust SQL query so the the result for missing table returns empty set. It can be used with `config.ignore-missing-values` command line parameter.

Version-specific queries work fine within a context of a single database engine, so it might be a bit tricky to come up with a generic and reliable solution. Overall, I might get back to it after v1.0 is released. Until then, we have to use one of the options above.

**Reply by blueeye2015** [reply]
Thank you for your replay ,understood

**Comment by keithf4** [comment]
For PostgreSQL, I've created views that call a function. That function contains the logic to check for which version is installed and runs the appropriate query. Makes it more complicated from the database end, but vastly simplifies it from the exporter end and is a lot more flexible long term. If another version change comes out, just update the function and everything continues to work as before without any other changes needed.

Pretty sure the same could be done for mysql

**Reply by burningalchemist** [reply]
@keithf4 Thanks for another solution! 🙂👍
